server/dbInit.ts

```diff
@@ -1,6 +1,165 @@
+ */
+export async function fetchClassesForSubject(endpoint: string, semester: string, subject: ScrapingSubject): Promise<ScrapingClass[]> {
+  const result = await axios.get(`${endpoint}search/classes.json?roster=${semester}&subject=${subject.value}`, { timeout: 10000 });
+  if (result.status !== 200 || result.data.status !== "success") {
+    console.log(`Error fetching subject ${semester}-${subject.value} classes! HTTP: ${result.statusText} SERV: ${result.data.status}`);
+    return null;
+  }
+
+  return result.data.data.classes;
+}
+
+/*
+ * Fetch the relevant classes, and add them to the collections
+ * Returns true on success, and false on failure.
+ */
+export async function fetchAddCourses(endpoint: string, semester: string): Promise<boolean> {
+  const subjects = await fetchSubjects(endpoint, semester);
+
+  const v1 = await Promise.all(subjects.map(async (subject) => {
+    const subjectIfExists = await Subjects.findOne({ subShort: subject.value.toUpperCase() }).exec();
+
+    if (!subjectIfExists) {
+      console.log(`Adding new subject: ${subject.value}`);
+      const res = await new Subjects({
+        _id: shortid.generate(),
+        subShort: subject.value.toUpperCase(),
+        subFull: subject.descrformal,
+      }).save().catch((err) => {
+        console.log(err);
+        return null;
+      });
+
+      // db operation was not successful
+      if (!res) {
+        throw new Error();
+      }
+    }
+
+    return true;
+  })).catch((err) => null);
+
+  if (!v1) {
+    console.log("Something went wrong while updating subjects!");
+    return false;
+  }
+
+  // Update the Classes in the db
+  const v2 = await Promise.all(subjects.map(async (subject) => {
+    const classes = await fetchClassesForSubject(endpoint, semester, subject);
+
+    // skip if something went wrong fetching classes
+    // it could be that there are not classes here (in tests, corresponds to FEDN)
+    if (!classes) {
+      return true;
+    }
+
+    // Update or add all the classes to the collection
+    const v = await Promise.all(classes.map(async (cl) => {
+      const classIfExists = await Classes.findOne({ classSub: cl.subject.toUpperCase(), classNum: cl.catalogNbr }).exec();
+
+      // The class does not exist yet, so we add it
+      if (!classIfExists) {
+        console.log(`Adding new class ${cl.subject} ${cl.catalogNbr}`);
+        const res = await new Classes({
+          _id: shortid.generate(),
+          classSub: cl.subject.toUpperCase(),
+          classNum: cl.catalogNbr,
+          classTitle: cl.titleLong,
+          classFull: `${cl.subject.toUpperCase()} ${cl.catalogNbr} ${cl.titleLong}`,
+          classSems: [semester],
+          // TODO: provide initial values here?
+          classProfessors: null,
+          classRating: null,
+          classWorkload: null,
+          classDifficulty: null,
+        }).save().catch((err) => {
+          console.log(err);
```

Comment:
Same idea as above, helpful to preempt the error you print, like this:

https://github.com/cornell-dti/course-reviews-react-2.0/blob/f30c4ca2f22682e574e47f703c3fd1100d558f6b/server/methods.ts#L198

Reply:
If this fails then the if at line 121 fails, and we get an error message about the particular class it was unable to insert.

Reply:
This is true, sounds good